File: client/mobile-app/utils/avatarApi.ts
// Avatar API client for React Native app
import { getApiClient } from './api';

export interface AvatarStatus {
  status: 'pending' | 'generating' | 'ready' | 'failed';
  progress?: number;
  error?: string;
  modelUrl?: string;
  textureUrls?: string[]; // Separate texture image URLs (extracted from GLB for React Native)
  animationUrls?: string[]; // Separate animation GLB URLs (for Meshy models)
  format?: string;
  modelType?: string;
  estimatedTimeRemaining?: number;
  binFileName?: string; // For GLTF format: the .bin filename referenced in the GLTF JSON
  binUrl?: string; // Explicit signed URL for the .bin file (needed for private containers)
}

export interface AvatarDownloadUrl {
  url: string;
  expiresIn: number | null;
  format: string;
  modelType: string;
  binFileName?: string; // For GLTF format: the .bin filename referenced in the GLTF JSON
  binUrl?: string; // Explicit signed URL for the .bin file (needed for private containers)
}

export interface Avatar {
  agentId: string;
  ownerUserId: string;
  modelType?: string;
  format?: string;
  modelUrl?: string;
  status: string;
  generationStartedAt?: string;
  generationCompletedAt?: string;
  generationError?: string;
  provider?: string;
  characterDescription?: any;
}

export const avatarApi = {
  /**
   * Get avatar for an agent
   */
  getAvatar: async (agentId: string): Promise<Avatar | null> => {
    const api = getApiClient();
    try {
      return await api.get<Avatar>(`/avatars/${agentId}`);
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  },

  /**
   * Get avatar generation status with progress
   */
  getAvatarStatus: async (agentId: string): Promise<AvatarStatus> => {
    const api = getApiClient();
    return api.get<AvatarStatus>(`/avatars/${agentId}/status`);
  },

  /**
   * Get signed download URL for avatar model
   * @param agentId Agent ID
   * @param expiresSeconds Optional expiration time in seconds (default: 900 = 15 minutes)
   */
  getDownloadUrl: async (agentId: string, expiresSeconds?: number): Promise<AvatarDownloadUrl> => {
    const api = getApiClient();
    const params = expiresSeconds ? `?expiresSeconds=${expiresSeconds}` : '';
    return api.get<AvatarDownloadUrl>(`/avatars/${agentId}/download-url${params}`);
  },

  /**
   * Verify that model files exist in storage
   * @param agentId Agent ID
   */
  verifyFiles: async (agentId: string): Promise<{
    modelUrl: string;
    modelExists: boolean | null;
    binExists: boolean | null;
    binFileName?: string;
    format?: string;
    error?: string;
  }> => {
    const api = getApiClient();
    return api.get(`/avatars/${agentId}/verify-files`);
  },

  /**
   * Start avatar generation for an agent
   * Note: This is usually triggered automatically when an agent is created
   */
  generateAvatar: async (agentId: string, agentProfile: any): Promise<{ agentId: string; status: string; message: string }> => {
    const api = getApiClient();
    return api.post<{ agentId: string; status: string; message: string }>('/avatars/generate', {
      agentId,
      agentProfile,
    });
  },
};

File: client/unity/ARAvatarClient/Assets/Scripts/API/AvatarAPI.cs
using System.Collections;
using UnityEngine;
using AIChatAR.Models;
using AIChatAR.API;

namespace AIChatAR.API
{
    /// <summary>
    /// Avatar API client - matches React Native avatarApi.ts
    /// </summary>
    public class AvatarAPI : MonoBehaviour
    {
        private APIClient apiClient;
        private string avatarServiceUrl;

        public void Initialize(string baseUrl, string authToken = null)
        {
            // Load config to get avatar service URL
            TextAsset configFile = Resources.Load<TextAsset>("Config");
            if (configFile != null)
            {
                Config config = JsonUtility.FromJson<Config>(configFile.text);
                avatarServiceUrl = config.avatarServiceUrl;
            }
            else
            {
                // Fallback to baseUrl + /avatars
                avatarServiceUrl = $"{baseUrl}/avatars";
            }

            apiClient = new APIClient(avatarServiceUrl, authToken);
        }

        /// <summary>
        /// Get avatar for an agent - matches getAvatar()
        /// </summary>
        public void GetAvatar(string agentId, System.Action<AIChatAR.Models.Avatar> onSuccess, System.Action<string> onError)
        {
            StartCoroutine(apiClient.Get<AIChatAR.Models.Avatar>(
                $"/{agentId}",
                (avatar) => {
                    if (avatar != null)
                    {
                        onSuccess?.Invoke(avatar);
                    }
                    else
                    {
                        onError?.Invoke("Avatar not found");
                    }
                },
                (error) => {
                    if (error == "Not found")
                    {
                        onSuccess?.Invoke(null); // Return null for 404, matching React Native behavior
                    }
                    else
                    {
                        onError?.Invoke(error);
                    }
                }
            ));
        }

        /// <summary>
        /// Get avatar generation status with progress - matches getAvatarStatus()
        /// </summary>
        public void GetAvatarStatus(string agentId, System.Action<AvatarStatus> onSuccess, System.Action<string> onError)
        {
            StartCoroutine(apiClient.Get<AvatarStatus>(
                $"/{agentId}/status",
                onSuccess,
                onError
            ));
        }

        /// <summary>
        /// Get signed download URL for avatar model - matches getDownloadUrl()
        /// </summary>
        public void GetDownloadUrl(string agentId, System.Action<AvatarDownloadUrl> onSuccess, System.Action<string> onError, int expiresSeconds = 900)
        {
            string endpoint = $"/{agentId}/download-url?expiresSeconds={expiresSeconds}";
            StartCoroutine(apiClient.Get<AvatarDownloadUrl>(
                endpoint,
                onSuccess,
                onError
            ));
        }

        /// <summary>
        /// Start avatar generation for an agent - matches generateAvatar()
        /// </summary>
        public void GenerateAvatar(string agentId, object agentProfile, System.Action<object> onSuccess, System.Action<string> onError)
        {
            var body = new
            {
                agentId = agentId,
                agentProfile = agentProfile
            };

            StartCoroutine(apiClient.Post<object>(
                "/generate",
                body,
                onSuccess,
                onError
            ));
        }
    }
}

